### DISCUSSION TITLE
Add Coin Selection Algos

### DISCUSSION THREAD

**Discussion by yancyribbens** [rust-bitcoin/rust-bitcoin#1402] [Ideas]
Bitcoin Core has a number of [coin-selection](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.cpp) strategies that are not currently represented here in rust-bitcoin.  I'd like to start a discussion about adding the coin-selection algos used by Bitcoin Core to Rust Bitcoin, in particular, BnB search.  To that end, I can offer an early implementation I began some time ago as a [starting point](https://github.com/p2pderivatives/rust-bitcoin-coin-selection).

The questions I have are:

1)  do we like the strategy implemented [here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection) as a starting point to iterate on.
2) what needs to be be changed/updated at a minimum to move this under the rust-bitcoin commons.
3) what are some improvements/recommendations that should be added going forward.  I have a list of improvements that I can think of, although I'm interested to hear what others have in mind.

Also note, there is a Rust coin-selection strategy in the BDK [here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection) that can also be used for inspiration and discussion.

**Comment by DanGould** [comment]
rust-bitcoin seems more focused on strong typing for bitcoin crypto & serialization. Wallet management including coin selection seems to be more of the role of BDK, where you'll find lots of discussion about this problem already.

After BDK 1.0 is released I hear rumors there will be a concerted effort to make coin selection a first class problem that BDK intends to solve. There are some musings here: https://github.com/bitcoindevkit/bdk/pull/773. @evanlinjin would know more.

**Reply by Kixunil** [reply]
Yeah, let's not make this library everything but a kitchen sink. A simple coin selection crate could live in this org but I'm not sure if it makes sense without something like BDK.

**Reply by yancyribbens** [reply]
@DanGould I noticed the `rust-wallet` hasn't implemented a sophisticated selection strategy yet (see the [TODO](https://github.com/rust-bitcoin/rust-wallet/blob/master/src/coins.rs#L284).  If we had a standalone crate for coin-selection that could be used as a dependency by the `rust-wallet`.  If BDK also develops something similar, maybe we could coordinate efforts or use whichever.  In the meantime, I don't mind advancing a version under the `rust-bitcoin` _organization_.

**Comment by apoelstra** [comment]
I agree that coin selection algorithms do not belong in rust-bitcoin the project. When we were talking on IRC @yancyribbens I thought you meant moving a coin selection crate into the rust-bitcoin **organization**, which I'd be fine with provided the crate met our requirements (mostly, being actively maintained and having a MSRV of 1.41).

**Reply by yancyribbens** [reply]
@apoelstra I verified this crate works with version `1.41` by changing the default version of `rustc` to `1.41` using `rustup default 1.41`.  I can also commit to being a active maintainer.  Just curious, why `1.41`?

**Reply by apoelstra** [reply]
We have three criteria that we use to choose our MSRV:

* It should be available in debian stable (and sometimes oldstable if stable is really new :))
* It should be (close to) the version compilable by mrustc, so that people can bootstrap the compiler in a reasonable time without downloading rustc binaries
* It should be >2 years old

We also have a soft requirement that we don't like changing the MSRV very frequently, because every time we have to do it across several crates at once and we're still unsure what the sanest process for that should be.

We have [this tracking issue](https://github.com/rust-bitcoin/rust-bitcoin/issues/1060#issuecomment-1308614955) where we note which compiler features we'd really like to have, so that when we **do** bump our MSRV we can try to choose the "most stable one" that has all the features we want.

We also have [this issue which proposes bumping the MSRV now to 1.48](https://github.com/rust-bitcoin/rust-bitcoin/discussions/1329) but it has not achieved consensus so I'm doubtful that it'll happen anytime soon.

**Comment by yancyribbens** [comment]
@apoelstra when you suggested starting a discussion to include coin-selection as part of the rust bitcoin org I started the discussion here.  Is there a place I should move this discussion so it's not confusing?  I did intend to include coin-selection as a stand-alone repo in rust-bitcoin organization.

**Reply by Kixunil** [reply]
You wrote at the correct place. The issue is we have a naming collision - Rust Bitcoin is both name of the organization and of the crate. I suggest explicitly typing "Rust Bitcoin organization" when you mean that. We don't have a better way to distinguish yet.

**Reply by yancyribbens** [reply]
Ok thanks.  I'll keep that in mind in the future!

**Reply by apoelstra** [reply]
Yep, sorry -- and it does add to the confusion that org-wide discussions happen in this specific repo, because AFAIK there is no better place for them.

**Comment by yancyribbens** [comment]
Hi all, I wanted to follow up on this discussion.  It seems to me that the main objection to adding a standalone coin-selection repo to the rust-bitcoin organization is the rumor that BDK may be also working on this (or working on this in the future?).   I think the other criteria have been met (active maintainer and compatible MSRV).  Is there any other feedback, comments or considerations?  If not, do we want to add this under the rust-bitcoin organization?

**Reply by apoelstra** [reply]
I think the actual blocker is that we didn't realize you were proposing to move a specific crate into this org :)

**Reply by sanket1729** [reply]
Is there a crate that we can look at? Perhaps if the crate is already up, we can offer specific feedback.

**Reply by yancyribbens** [reply]
@apoelstra @sanket1729 sorry for the conusion - yes I'm proposing adding a crate for coin-selection to this organization.

The original crate to be move is here - https://github.com/p2pderivatives/rust-bitcoin-coin-selection
I also have a fork which is a few commits ahead currently here - https://github.com/yancyribbens/rust-bitcoin-coin-selection 

Any specific feedback is appreciated.

**Reply by sanket1729** [reply]
Is it just 150 lines of non-test code? Should be easy to review and move in here. Superficially, this looks good. 

From my side, It would be nice to README and a basic CI setup to test MSRV.

**Reply by sanket1729** [reply]
I might review the code in more detail later.

**Reply by yancyribbens** [reply]
@sanket1729 Yes, it's only BnB currently and is small but has good test coverage imo.  I can add a README and basic CI setup before adding it to the organization.  Is there an example or documentation for the CI setup you can recommend?

**Reply by Kixunil** [reply]
Oh, wow, from quick glance it looks really simple. I'm just not sure about the `Utxo` trait. Maybe it could be just `Borrow<TxOut>` (but I'm very tired rn so I could be totally missing something)

**Reply by tcharding** [reply]
For CI we use bash scripts, they are all slightly different, and a total time sink - feel free to ping me if you have any trouble. My suggestion would be to write a minimal one, we can iterate later (see `rust-bitcoin/bitcoin/contrib/test.sh` and `rust-bitcoin/hashes/contrib/test.sh` for examples). The CI scripts are run by `rust.yml` github action but you should not need to touch that).

**Reply by yancyribbens** [reply]
@tcharding thanks.  I think this should do for a minimal one to start: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/e172a7830c2c5b3f13700e80fc625bb0a54058b9

**Reply by tcharding** [reply]
Thanks for having a go man, I added some comments to that link.

**Comment by tcharding** [comment]
I added issue https://github.com/p2pderivatives/rust-bitcoin-coin-selection/issues/9 with a few observations.

**Reply by yancyribbens** [reply]
@tcharding thanks!  There are a few PR's to address your observations:

https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/8
https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/10

**Reply by yancyribbens** [reply]
@tcharding thanks for all the help getting this ready.  I've added one last change for now to test.sh which is bechmarks: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/15

**Comment by murchandamus** [comment]
I glanced at the Branch and Bound algorithm in the https://github.com/p2pderivatives/rust-bitcoin-coin-selection repository. It seems to me that there might be some basic features missing before it should be merged into a production system. For example, I have the impression that it currently runs with the _values_ of the UTXOs rather than the _effective values_. This means that the input selection will not account for the fees that the inputs need, it will be undershooting the target.

I will also leave some issues on the repository to point out some additional observations.

**Reply by apoelstra** [reply]
Thanks for taking a look!

It looks like these issues are serious enough that the BnB algorithm won't actually hit the target feerate -- which is serious but fortunately a very easy property to measure, which I hadn't considred until now.

Before bringing this crate into rust-bitcoin I'd like to see a fuzztest that attempts to produce various transactions with various feerates, and guarantees that we always hit the target (or fail to select).

**Reply by yancyribbens** [reply]
Thanks @Xekyo @apoelstra  - It's true this current implementation doesn't consider the fee rate, which seems to be a large oversight.  In fact, in working on some further performance optimizations recently, I believe I've found other edge cases where a solution should be found but in fact is not (another serious problem).  I'll create this as an issue to be solved as well.

**Reply by murchandamus** [reply]
I find that first trying to achieve the main objectives  first and using that to building up a bit of a test corset can help a lot with ensuring that refactors for optimizations do not introduce bugs. Often, the code that just solves the main objectives is also fast enough, and it is unnecessary to complicate the code to speed it up. E.g. when people only call coin selection once an hour, it does not matter that much whether it takes 0.5 s or 0.1 s. So, perhaps performance improvements should only become the focus when a user of the crate complains that they are using it to build tens of thousands of transactions per day and it is not fast enough. :)

**Reply by yancyribbens** [reply]
Good piont.  And my recent contributions to core have helped my understanding of the logic (and waste metric) much more since I wrote this first years ago.

**Comment by yancyribbens** [comment]
I'm a bit disappointed that I didn't catch this earlier, however it's in part because the tests used worked primarily with whole numbers which I pulled from the current core implementation.  If more tests with fractional inputs are considered, there are combinations of UTXO that should be found but are not.  Here's the issue I created:

https://github.com/p2pderivatives/rust-bitcoin-coin-selection/issues/20

This will probably require a complete re-write taking into account some of @Xekyo findings as well.

**Reply by yancyribbens** [reply]
@Xekyo I created a PR to fix missing solution problems.  Before I move on to fixing your suggestions, I wondered if you wanted to a take a look.  It seems like before implementing back tracking, the next thing to do is work on the waste calculation.  Thanks!

**Reply by yancyribbens** [reply]
https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/21

**Reply by yancyribbens** [reply]
SRD selection algorithm PR: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22

**Comment by yancyribbens** [comment]
There's a new PR for SRD here: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/23

**Comment by yancyribbens** [comment]
does anyone else have any suggestions before merging this? https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/23

**Comment by stevenroose** [comment]
I was actually looking into coin selection recently and eventually ended up piggy backing BDK's implementation. I reached out to them about the idea of pulling some non-BDK-specific wallet-related things into the bitcoin-wallet crate that is archived now. It was kinda annoying having to use the BDK traits in order to use their impl. It could easily be generalized.

But this new crate seems interesting.

**Reply by yancyribbens** [reply]
Well, it's not really new.  I've been working on it on and off for an embarrassingly long time.  I'm a bit stalled at the moment but maybe taking a look at the BDK implementation will help.

**Comment by yancyribbens** [comment]
I've merged an updated version of `SRD` https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/23.  Thanks again Murch (http://murch.one/) and Thibaut Le Guilly for the reviews.

**Comment by yancyribbens** [comment]
Happy New Years all,

I've created a new Rust BnB coin selection algo here: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/28

Also of interest, I did a quick benchmark to compare this new Rust implementation vs the existing Bitcoin Core (timed on my laptop):

Rust Implementation: `897,593 ns/op`

C++ implementation: `816,374 ns/op`

Maybe there's some Rust compiler tricks to improve the Rust implementation further? Appreciate any feedback or review.

Cheers

**Comment by yancyribbens** [comment]
I've created a new PR to change the return type of Single Random Draw: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/33

cc @Kixunil

**Comment by yancyribbens** [comment]
I think this interface is ready for a V1 version and can be published:

interface in `lib.rs`
```rust
pub fn select_coins(
    target: Amount,
    cost_of_change: Amount,
    fee_rate: FeeRate,
    long_term_fee_rate: FeeRate,
    weighted_utxos: &[WeightedUtxo],
) -> Option<std::vec::IntoIter<&WeightedUtxo>> {
```
latest PR to cleanup interface: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/36

@apoelstra @Kixunil does this interface seem sane?  Is there any other items that should be complete before moving this over to the rust-bitcoin organization?

@murchandamus friendly ping that https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/30 is still open for review.  I believe this PR closes all the issues you opened.

Cheers!

**Reply by Kixunil** [reply]
I think rather than `cost_of_change` it should accept weight prediction for the change output and compute the cost internally. And perhaps make UTXO generic (at least it could be `AsRef<WeightedUtxo>`).

**Reply by yancyribbens** [reply]
I would like to make UTXOs generic also, and I think that can be done easy enough using weight prediction instead of WeightedUtxo.  Getting rid of `cost_of_change` seems like a good idea too, although it has the downside of not maintaining parity with the core implementation, so i'm more ambivalent about making that change for v1.  I'll wait a little bit probably before rushing into any more changes as it would be nice to get some more feedback if possible.

**Reply by Kixunil** [reply]
> core implementation

Not a library, so it doesn't matter. Libraries should aim at being most useful and if the value can be computed using `FeeRate` it makes more sense to not force this computation on the caller.

**Reply by yancyribbens** [reply]
@Kixunil I've been thinking about your suggestions.  What about something like this:
```
pub fn select_coins(
    target: Amount,
    fee_rate: FeeRate,
    long_term_fee_rate: FeeRate,
    change_output_prediction: InputWeightPredition,
    utxos: &[(i: InputWeightPrediction, utxo_amount: Amount)],
) -> Option<std::vec::IntoIter<&WeightedUtxo>> {
```

Remove cost_of_change replacing with `change_output_prediction`.  Also, change utxos to be a tuple where the first item is the input_weight_prediction of the TxOut to possibly spend and the second item is the utxo amount (value).  Maybe this could instead be ExtendedTxIn come to think of it?  I'm not sure if ExtendedTxIn can compute the weight prediction.

**Reply by Kixunil** [reply]
output is not an input. I guess you'll see better what I mean when I get around making the changes.

**Reply by yancyribbens** [reply]
> output is not an input. I guess you'll see better what I mean when I get around making the changes.

I had thought ExtendedTxIn had a way to reference the output that created it.  Just thinking about a way for the Output to predict it's size instead of using WeightedUtxo.

**Reply by Kixunil** [reply]
That's a previous one, change is the new one.

**Reply by yancyribbens** [reply]
@Kixunil I added a PR to use InputWeightPrediction instead of cost_of_change: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/37

**Reply by yancyribbens** [reply]
> And perhaps make UTXO generic (at least it could be AsRef<WeightedUtxo>)

@Kixunil What about this?:
```rust
pub fn select_coins<T: WeightedUtxo>(
    target: Amount,
    fee_rate: FeeRate,
    change_weight_prediction: InputWeightPrediction,
    long_term_fee_rate: FeeRate,
    weighted_utxos: &[T],
) -> Option<std::vec::IntoIter<&T>> {
```

And the trait used in the trait bound:
```rust
/// The WeightedUtxo trait provides the necessary interfaces for coin-selection. 
pub trait WeightedUtxo {
    /// The weight of a UTXO once the spending conditions are satisfied.
    fn satisfaction_weight(&self) -> Weight;
    /// The concrete UTXO.
    fn utxo(&self) -> TxOut;
}
```

**Reply by yancyribbens** [reply]
> And perhaps make UTXO generic (at least it could be AsRef<WeightedUtxo>).

@Kixunil I finally got around to making UTXO generic: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/58

I needed this so that I could implement Arbitrary for fuzzing without needing to add Arbitrary as a dependency to `rust-bitcoin-coin-selection`.

**Reply by yancyribbens** [reply]
> And perhaps make UTXO generic (at least it could be AsRef<WeightedUtxo>).

I think I'm going to remove the generic type parameters because it's painting me into a corner.  Recently I added a min-heap to single_random_draw which is used to minimize the selection size which requires Ord.  However, there doesn't seem to be a good way enforce how Ord is implemented on a foreign type.  Trying to define Ord for a trait doesn't seem to hold if the foreign type implements Ord a different way.  And it seems annoying to push that onto the consumer that they implement Ord and implement it correctly.

**Comment by yancyribbens** [comment]
> Before bringing this crate into rust-bitcoin I'd like to see a fuzztest that attempts to produce various transactions with various feerates, and guarantees that we always hit the target (or fail to select).

@apoelstra is this still something you would like to see done?  @murchandamus was able to give it a review but I think there might be some confusion because the project is in Rust (i'm still trying to get some clarification).  I believe fuzz testing or model testing could be added to show all solutions are found.  The other disagreement is if this is actually a depth first search which I think can be shown that it is.  Part of the issue I'm running into is that there's not really anybody that seems to know bitcoin cores implementation well enough to code review and that also has the time/knowledge of rust to offer feedback.  I guess in short - any help/advice here would be greatly appreciated.

**Reply by apoelstra** [reply]
At a minimum, yes, I would like to see fuzztesting that the code is never wrong in the "bad" direction.

But I would also like to see a reviewer signoff from somebody known and active in the rust-bitcoin community, and as you say, reviewer bandwidth is extremely limited which means that I don't know who or when will find time.

**Reply by yancyribbens** [reply]
Alright, thanks for the feedback.  I'll add fuzz testing then to my todo list.

**Reply by yancyribbens** [reply]
> At a minimum, yes, I would like to see fuzztesting that the code is never wrong in the "bad" direction.

@apoelstra I added honggfuzz testing for both SRD and BNB https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/59.  Since Arbitrary is not yet in rust-bitcoin I am using a git branch of rust-bitcoin, although I was thinking of finishing https://github.com/rust-bitcoin/rust-bitcoin/pull/3015 next.  Also I think we could improve the fuzz tests in rust-bitcoin using Arbitrary as well.

**Comment by yancyribbens** [comment]
I'm going to go ahead and merge https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/30 since it's unlikely I'll get anymore feedback.  I think it's 100x better than the previous implementation.  Also the benchmark times are comparable to the bitcoin core implementation.  Furthermore, I'm convinced after the review, feedback and addition of more tests, that the implementation results are correct.  Thanks to all those that provided reviews so far, it's very much appreciated!

**Reply by yancyribbens** [reply]
Merged.

**Comment by yancyribbens** [comment]
Hi all,

Sorry it's taken a while to get back to this.  I wanted to re-visit adding this to rust-bitcoin with the latest PR adding proptests: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/63 which shows the correctness of the provided algorithms SRD and BnB.  Since the last time I've posted, I've made a number of improvements:

* Added proptests using Arbtest
* Added fuzz test targets using Libfuzzer
* Refactored BnB and SRD test suit to simplify tests
* Added unit tests to the lib module
* Fixed some minor bugs I uncovered while fuzz testing and proptesting.

It would be great if I could find someone from rust-bitcoin to review this again.  In the past @Kixunil reviewed the project so he is already familiar somewhat with the topic area.

cc @apoelstra @murchandamus

**Reply by yancyribbens** [reply]
fyi I just added a test case that was uncovered by the proptests: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/63/commits/23f5186a98c590064907143f1c4b6dce897532b6 and updated the proptest framework accordingly.  The proptest had failed because it created a UTXO greater than MAX_MONEY which BnB then didn't find a solution for (which is correct since a Utxo greater than MAX_MONEY should not have a solution).  I have these tests running locally 24/7 so it may uncover more weird edge cases.

**Reply by yancyribbens** [reply]
@apoelstra we've been in correspondence about moving this to rust-bitcoin over the last few years.  I'm still interested in maintaining this under the rust-bitcoin org and I believe the last round of updates including fuzzing and proptesting meets your requested changes.  I'm afraid with @Kixunil MIA it might be difficult to find someone to review this, although @murchandamus has provided more than sufficient review and I don't want to pester him further.  Friendly request for further feedback on porting this to rust-bitcoin.  Cheers.

**Reply by yancyribbens** [reply]
@apoelstra sorry to pester you about this, although I noticed in a related thread that you are the designated person who publishes all crates in rust-bitcoin.  That reminds me that I had published an unstable version of `bitcoin-coin-selection` on crates.io a while back when there was discussion about name squatting and names not available, so I thought I would be pro-active and publish this before that happened.  The unstable version published has received adequate review (thanks again @murchandamus ), and since then I've added much more testing to show this.  Anyway, I am not against changing what's on crates.io, although, as it stands, the `bitcoin-coin-selection` links are currently pointing to rust-bitcoin organization because of our discussion here to move the crate.  If that's not going to happen in the near future, then I should probably update those links in the meantime.

Also, I have some things I'd like to do to further cleanup the crate, like make the different selection algorithms traits that are implemented on a UtxoPool struct or TxOutPool.  Maybe UtxoPool could be a struct that rust-bitcoin defines?  Although I'm not sure if that's useful outside of coin-selection.  And there's still coin-grinder to add which is a different type of branch and bound search.

Cheers

**Reply by apoelstra** [reply]
I don't think I'm the designated publisher in general. For chacha20-poly1305 it makes sense for me to do it since I care about that crate, have read the code, and am willing to co-maintain it. So I didn't argue there. But I don't think it makes sense as a general rule. (Though I suppose I should have publishing *rights* for everything in the org. For example, Kix is the "designated publisher" for hex-conservative but in his absense I released 0.3.0.)

This crate is harder. To move into the org somebody needs to take a careful look at it. It sounds like only Kix cares enough about coin selection to dive into this. So you may be stuck waiting for him to come back. I can't blindly bring it in, and it will be a long/indefinite amount of time before I can review it.

**Reply by yancyribbens** [reply]
That makes sense, and maybe this is a good time to shift my focus to other projects for a while.  It's been enjoyable to work on, and there's still so much more to continue with.  Although it's difficult at times without much outside involvement or direction.  At any rate, I'm happy with the progress that has been made over the last few years.  And if you or @Kixunil has any feedback or ideas I'm happy to continue working together.

**Reply by yancyribbens** [reply]
@apoelstra I was thinking, maybe there's some issues that would languish that you would like to see complete that I will work on in exchange for your help in reviewing this to bring it over?  For example, maybe switching over to libfuzzer, add Proptests to Amount , or  [Check integer types implement correct set of](https://github.com/rust-bitcoin/rust-bitcoin/issues/1639) as a few that come to mind.  Not to make this super transactional, although there is more than just @Kixunil who "cares" about this, and there's a real shortage of maintainers active here (now just you and Tobin) that I can work with on this.  Just an idea.

**Reply by Kixunil** [reply]
> It sounds like only Kix cares enough about coin selection to dive into this.

Sadly, "cares" is a strong word. It was just curiosity. I might revisit it if I ever decide to create my own wallet (which I was thinking about for a bit) but so far, I don't think I'll find the time to do anything significant with UTXO selection. Though I wonder what BDK does and maybe they would be more into it?

**Reply by yancyribbens** [reply]
> Though I wonder what BDK does and maybe they would be more into it?

BDK rolls their own selection algorithms last time I took a look, and only BnB and SRD I think.  Also, I don't think they include benchmarks and some of the other testing frameworks I've added.

> I might revisit it if I ever decide to create my own wallet (which I was thinking about for a bit) but so far, I don't think I'll find the time to do anything significant with UTXO selection

That's fair.  I've been thinking that would be my next step as well, time permitting.

**Comment by yancyribbens** [comment]
Added UTXO exclusion shortcut: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/67

**Comment by yancyribbens** [comment]
I've recently added [coin-grinder](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/75/commits/fb02a5265108284f377d7c0e43abe55bac0c526a), a Rust DFS-based selection algorithm which optimizes for transaction
weight and creates a change output.  Any and all feedback is appreciated!

**Comment by yancyribbens** [comment]
I feel like this discussion has run it's course.  Thank you kindly for all the feedback.  I will continue to publish updates and improvements to Crates.io.
